### PR TITLE
elasticsearch 7.6

### PR DESCRIPTION
**Pull request by ankane**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Elasticsearch 7

A few notes:

- Builds the OSS flavor from source as requested in #38966 (the current Elasticsearch formulas don't build from source - wasn't able to tell if there was a reason for that)
- Uses 7.6.2. 7.7.0 is the latest, but it requires Java 14 to build from source
- The build script (`gradlew`) downloads gradle 3.6. I tried with the Homebrew version but received an error I didn't totally understand.
- `elasticsearch-translog` was replaced with `elasticsearch-shard` in ES 7 https://github.com/elastic/elasticsearch/pull/33373

```txt
An exception occurred applying plugin request [id: 'nebula.ospackage-base', version: '8.1.0']
> Failed to apply plugin [id 'nebula.ospackage-base']
   > Could not create an instance of type com.netflix.gradle.plugins.packaging.ProjectPackagingExtension.
      > Could not find matching constructor for: org.gradle.api.internal.file.copy.DefaultCopySpec(org.gradle.api.internal.file.BaseDirFileResolver, org.gradle.api.internal.file.DefaultFileCollectionFactory, org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator)
```

Follow up steps:

1. Update Kibana
2. Alias as `elasticsearch`

### CODE REVIEW

File: Formula/elasticsearch@7.6.rb [hunk 1]
```diff
@@ -0,0 +1,150 @@
+class ElasticsearchAT76 < Formula
+  desc "Distributed search & analytics engine"
+  homepage "https://www.elastic.co/products/elasticsearch"
+  url "https://github.com/elastic/elasticsearch/archive/v7.6.2.tar.gz"
+  sha256 "6ff4871dcae6954e13680aefc196da574a59a36418d06a7e095550ce81a370f8"
+
+  head do
+    url "https://github.com/elasticsearch/elasticsearch.git"
+  end
```
Comment (SMillerDev) [inline Formula/elasticsearch@7.6.rb:9]:
```suggestion
  sha256 "6ff4871dcae6954e13680aefc196da574a59a36418d06a7e095550ce81a370f8"
  head "https://github.com/elasticsearch/elasticsearch.git"
```

File: Formula/elasticsearch@7.6.rb [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+  depends_on "openjdk"
+
+  def cluster_name
+    "elasticsearch_#{ENV["USER"]}"
+  end
+
+  def install
+    # Build the package from source
```
Comment (SMillerDev) [inline Formula/elasticsearch@7.6.rb:18]:
```suggestion
    # Doesn't support brewed gradle
```

File: Formula/elasticsearch@7.6.rb [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      # 2. Configure paths
+      s.sub!(%r{#\s*path\.data: /path/to.+$}, "path.data: #{var}/lib/elasticsearch/")
+      s.sub!(%r{#\s*path\.logs: /path/to.+$}, "path.logs: #{var}/log/elasticsearch/")
+    end
+
+    inreplace "#{libexec}/config/jvm.options", %r{logs/gc.log}, "#{var}/log/elasticsearch/gc.log"
+
+    # Move config files into etc
+    (etc/"elasticsearch").install Dir[libexec/"config/*"]
+    (libexec/"config").rmtree
```
Comment (SMillerDev) [inline Formula/elasticsearch@7.6.rb:48]: Can you do this before the changing of the config?

File: Formula/elasticsearch@7.6.rb [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+  def post_install
+    # Make sure runtime directories exist
+    (var/"lib/elasticsearch").mkpath
+    (var/"log/elasticsearch").mkpath
+    ln_s etc/"elasticsearch", libexec/"config" unless (libexec/"config").exist?
+    (var/"elasticsearch/plugins").mkpath
+    ln_s var/"elasticsearch/plugins", libexec/"plugins" unless (libexec/"plugins").exist?
+    # fix test not being able to create keystore because of sandbox permissions
+    system bin/"elasticsearch-keystore", "create" unless (etc/"elasticsearch/elasticsearch.keystore").exist?
```
Comment (SMillerDev) [inline Formula/elasticsearch@7.6.rb:65]: Is this also useful for users? Or just for the test?

File: Formula/elasticsearch@7.6.rb [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+          <key>StandardOutPath</key>
+          <string>#{var}/log/elasticsearch.log</string>
+        </dict>
+      </plist>
+    EOS
+  end
+
+  test do
+    port = free_port
+    system "#{bin}/elasticsearch-plugin", "list"
```
Comment (SMillerDev) [inline Formula/elasticsearch@7.6.rb:113]: What is this for?

File: Formula/elasticsearch@7.6.rb [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    port = free_port
+    (testpath/"config/elasticsearch.yml").write <<~EOS
+      path.data: #{testpath}/data
+      path.logs: #{testpath}/logs
+      node.name: test-es-path-conf
+      http.port: #{port}
+    EOS
+
+    cp etc/"elasticsearch/jvm.options", "config"
+    cp etc/"elasticsearch/log4j2.properties", "config"
```
Comment (SMillerDev) [inline Formula/elasticsearch@7.6.rb:136]: Why are these needed?

File: Formula/elasticsearch@7.6.rb [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    pid = testpath/"pid"
+    begin
+      system "#{bin}/elasticsearch", "-d", "-p", pid
+      sleep 10
+      system "curl", "-XGET", "localhost:#{port}/"
+      output = shell_output("curl -s -XGET localhost:#{port}/_cat/nodes")
+      assert_match "test-es-path-conf", output
+    ensure
+      Process.kill(9, pid.read.to_i)
+    end
```
Comment (SMillerDev) [inline Formula/elasticsearch@7.6.rb:148]: I can't tell the difference between this and the other one. What are you testing additionally here?

**Review by SMillerDev — COMMENTED** [review]
@chrmoritz can you also take a look at this?

File: Formula/elasticsearch@7.6.rb [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    # Build the package from source
+    system "./gradlew", ":distribution:archives:oss-no-jdk-darwin-tar:assemble"
+
+    mkdir "tar" do
+      # Extract the package to the tar directory
+      system "tar", "--strip-components=1", "-xf",
+        Dir["../distribution/archives/oss-no-jdk-darwin-tar/build/distributions/elasticsearch-oss-*.tar.gz"].first
+
+      # Install into package directory
+      libexec.install "bin", "config", "lib", "modules"
```
Comment (chrmoritz) [inline Formula/elasticsearch@7.6.rb:27]:
```suggestion
      libexec.install "bin", "lib", "modules"
      (etc/"elasticsearch").install Dir["config/*"]
```

and updating the config `inreplace`'s below

File: Formula/elasticsearch@7.6.rb (hunk 4 above)
Comment (chrmoritz) [reply Formula/elasticsearch@7.6.rb:65]: A far as I understand this, we need a created keystore to run the tests. But because `etc/elasticsearch` is protected by the sandbox during test time, we need to create said keystore in `post_install`.

However I'm not sure how good and bad already creating a default keystore is for actual users of this formula.

**Review by chrmoritz — COMMENTED** [review]
Except updating `elasticsearch` and `kibana` to 7.x.x we need to find a strategy for what we will do with the `*beats` formulae. (also see: https://github.com/chrmoritz/homebrew-core/commits/elastic7).

Currently the new mage build system, which allows us to do a OSS only build, is only available for `auditbeat`, `filebeat` and `metricbeat`, however the following beats are **not** migrated yet: `heartbeat` and `packetbeat`.

We could either complete drop / deprecate these 2 beats or we would end up in a situation, where these 2 beats are only available in their versioned `@6` variants, but not for the latest 7.x.x version.

**Comment by ankane** [comment]
Since there was a lot of feedback, I just pushed against the `elasticsearch` formula to make discussion easier. Most of this is from the previous formula.

File: Formula/elasticsearch@7.6.rb (hunk 7 above)
Comment (ankane) [reply Formula/elasticsearch@7.6.rb]: The 2nd one tests a config file (not sure if two are really needed, but it's in the current version).

File: Formula/elasticsearch@7.6.rb (hunk 2 above)
Comment (ankane) [reply Formula/elasticsearch@7.6.rb]: Updated comment

File: Formula/elasticsearch@7.6.rb (hunk 1 above)
Comment (ankane) [reply Formula/elasticsearch@7.6.rb]: Removed `head` (originally left it since it was in current `elasticsearch` formula)

**Comment by SMillerDev** [comment]
LGTM :heavy_check_mark:

**Comment by ankane** [comment]
Thanks @SMillerDev. Have a minor change to update config before it's copied to `etc/elasticsearch` (as I'm not sure if the current approach will cause problems if `etc/elasticsearch` already exists). Will push that shortly.

**Comment by SMillerDev** [comment]
@chrmoritz Can you check again? 

Regarding the beats, I'm fine just updating the beats that are available as OSS. It's not something we can help.

**Review by SMillerDev — APPROVED** [review]
Thanks @ankane ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Comment by ankane** [comment]
Thanks @chrmoritz and @SMillerDev! Once #51883 is ready, we can upgrade Elasticsearch and Kibana to 7.7.x.
